### PR TITLE
Document the 'checks' module

### LINKED ISSUE #3331: embedded checks
**Related dev. issue(s):** https://github.com/tarantool/tarantool/issues/7726

**Product:** Tarantool
**Since:** 2.11
**Root document:** https://www.tarantool.io/en/doc/latest/reference/reference_lua/
**SME:** @ DifferentialOrange

# Details
Now tarantool has checks module on its board. checks is a lua module
previously distributed as a separate rock which is widely used by
many other tarantool lua modules (like cartridge, metrics and crud) and
tarantool applications. checks has its own repo with README covering its
API usage: https://github.com/tarantool/checks/blob/master/README.md .
Requested by @ DifferentialOrange in https://github.com/tarantool/tarantool/commit/5b68b6d624060cbee40c684ad608a03e1b290682.

### PR DESCRIPTION
**Pull request by andreyaksenov**
Added a new page for the ``checks`` module:
https://docs.d.tarantool.io/en/doc/embedded-checks/reference/reference_lua/checks/

> Note that new documentation topics now references testable code snippets from the [code_snippets](https://github.com/tarantool/doc/tree/latest/doc/code_snippets) folder.

### CODE REVIEW

File: doc/reference/reference_lua/checks.rst [hunk 1]
```diff
@@ -0,0 +1,506 @@
+..  checks-module:
+
+Module checks
+=============
+
+**Since:** :doc:`2.11.0 </release/2.11.0>`
```
Comment (DifferentialOrange) [inline doc/reference/reference_lua/checks.rst:6]: 1. It is available as a separate rock for all relevant Tarantool versions (1.10+)
2. Embedded checks package is provided with core Tarantool since 2.11.**1**

File: doc/reference/reference_lua/checks.rst [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+        *   -   **Functions**
+            -
+
+        *   -   :ref:`checks <checks-checks>`
+            -   When called inside a function, checks that the function's arguments conform to the specified types
+
+        *   -   **Objects**
+            -
+
+        *   -   :ref:`checkers <checks-checkers>`
```
Comment (DifferentialOrange) [inline doc/reference/reference_lua/checks.rst:310]: `checkers` is a global Lua variable (available disregarding whether one had called `require('checks')` in script or not). It's not clear from this documentation page.

File: doc/reference/reference_lua/checks.rst [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+
+..  class:: checkers
+
+    Checkers for different types.
+    This object provides access to checkers for Tarantool-specific types.
+    You can also add a :ref:`custom checker <checks_custom_function>` that performs arbitrary validations.
+
+    .. _checks-checkers-datetime:
+
+    .. function:: datetime(value)
```
Comment (DifferentialOrange) [inline doc/reference/reference_lua/checks.rst]: It is compiled like `checkers:datetime` when it's actually `checkers.datetime`

File: doc/reference/reference_lua/checks.rst [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+..  function:: checks(type_1, ...)
+
+    When called inside a function, checks that the function's arguments conform to the specified types.
+
+    :param string/table type_1: a :ref:`string <checks_string_type_qualifier>` or :ref:`table <checks_table_type_qualifiers>` type qualifier used to check the argument type
+    :param ...: optional type qualifiers used to check the types of :ref:`other arguments <checks_number_of_arguments>`
+
+..  _checks-checkers:
+
+checkers
```
Comment (DifferentialOrange) [inline doc/reference/reference_lua/checks.rst:328]: The only reason that `checkers` is public is that so user may add its own custom checkers. Existing checkers may be used as components for custom checkers, but there is no any other reason for one to call them explicitly. Maybe we should say that one shouldn't use `assert(checkers.datetime(arg))` instead of `checks('datetime')`

File: doc/reference/reference_lua/checks.rst (hunk 1 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst:6]: > It is available as a separate rock for all relevant Tarantool versions (1.10+)

Added a note after the introduction: https://docs.d.tarantool.io/en/doc/embedded-checks/reference/reference_lua/checks/.

> Embedded checks package is provided with core Tarantool since 2.11.1

Could you please clarify what do you mean by `Embedded checks package`? I have 2.11.0 on my machine and I can use `checks` without loading it explicitly:

```
tarantool
Tarantool 2.11.0-0-g247a9a4
type 'help' for interactive help
tarantool> checkers.
checkers.datetime(  checkers.error(     checkers.interval(  checkers.uint64(    checkers.uuid_bin(  
checkers.decimal(   checkers.int64(     checkers.tuple(     checkers.uuid(      checkers.uuid_str(  
```

File: doc/reference/reference_lua/checks.rst (hunk 2 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst:310]: Added a separate section about loading the module: https://docs.d.tarantool.io/en/doc/embedded-checks/reference/reference_lua/checks/#loading-checks

File: doc/reference/reference_lua/checks.rst (hunk 3 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst]: Fixed: https://docs.d.tarantool.io/en/doc/embedded-checks/reference/reference_lua/checks/#checkers

File: doc/reference/reference_lua/checks.rst (hunk 4 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst:328]: Updated the description to clarify that its main purpose is adding a custom checker: https://docs.d.tarantool.io/en/doc/embedded-checks/reference/reference_lua/checks/#checkers

File: doc/reference/reference_lua/checks.rst (hunk 1 above)
Comment (DifferentialOrange) [reply doc/reference/reference_lua/checks.rst:6]: Oh, it seems that I was wrong and this one was cherry-picked to 2.11.0. I had thought that the product approach would be the same as with `metrics`, which will appear only in 2.11.1.

File: doc/reference/reference_lua/checks.rst (hunk 2 above)
Comment (DifferentialOrange) [reply doc/reference/reference_lua/checks.rst:310]: Well, let's mention `checkers` in this sentence too.

Comment (DifferentialOrange) [reply doc/reference/reference_lua/checks.rst:310]: (To be honest, I had forgot that `checks` are global too.)

**Review by DifferentialOrange — APPROVED** [review]
LGTM from my side.

The documentation seems really nice and useful. It would be good to add a documentation page link to the [`checks`](https://github.com/tarantool/checks) repo README after merge.

**Comment by andreyaksenov** [comment]
> LGTM from my side.
> 
> The documentation seems really nice and useful. It would be good to add a documentation page link to the [`checks`](https://github.com/tarantool/checks) repo README after merge.

Thanks, will make a PR after merging.

File: doc/reference/reference_lua/checks.rst [hunk 5]
```diff
@@ -0,0 +1,516 @@
+..  checks-module:
+
+Module checks
+=============
+
+**Since:** :doc:`2.11.0 </release/2.11.0>`
+
+The ``checks`` module provides the ability to check the types of arguments passed to a Lua function.
+You need to call the :ref:`checks(type_1, ...) <checks-checks>` function inside the target Lua function and pass :ref:`one or more <checks_number_of_arguments>` type qualifiers to check the corresponding argument types.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst:9]: Nit: I'd start with "to check a function's argument types, call ..." to be more goal-oriented. In the current wording, the path to reader's goal is too long.

File: doc/reference/reference_lua/checks.rst [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+There are two types of type qualifiers:
+
+-   A :ref:`string type qualifier <checks_string_type_qualifier>` checks whether a function's argument conforms to the specified type.
+-   A :ref:`table type qualifier <checks_table_type_qualifiers>` checks whether the values of a table passed as an argument conform to the specified types.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: Maybe add short examples directly in the text?

File: doc/reference/reference_lua/checks.rst [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+.. NOTE::
+
+    For earlier versions, you can install the ``checks`` module from the `Tarantool rocks repository <https://www.tarantool.io/en/download/rocks>`_.
+
+.. _checks_loading:
+
+Loading checks
+--------------
+
+The :ref:`checks API <checks_api_reference>` is available in a script without loading the module using ``require('checks')``.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: Maybe say that `require` is needed only if the module is installed from repo?

I actually don't get how our devs decide which modules need `require` and which don't, but the explicit sentence "this module doesn't need require" looks a bit awkward.

File: doc/reference/reference_lua/checks.rst [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+
+
+.. _checks_number_of_arguments:
+
+Number of arguments to check
+----------------------------
+
+Depending on the number of arguments to be checked, you can pass one or more type qualifiers to the :ref:`checks(type_1, ...) <checks-checks>` function.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: The sentence reads a bit too formal.
Maybe smth simpler, like: "For each argument to check, you need to specify its own type qualifier"?

File: doc/reference/reference_lua/checks.rst [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+
+To check the types of several arguments, you need to pass the corresponding type qualifiers to the ``checks`` function.
+In the example below, both arguments should be string values.
+
+..  literalinclude:: /code_snippets/test/checks/checks_string_multiple_test.lua
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+To turn off checking of specific arguments, use the :ref:`? placeholder <checks_any_type>`.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: Do I get it right that `?` is for checking arbitrary subsets of arguments? For example, in `f(a, b, c, d)`, check only `a` and `d`? If yes, then maybe show an example?

Comment (p7nov) [reply doc/reference/reference_lua/checks.rst]: Also, `skip checking` seems better to me here than `turn off checking`.

File: doc/reference/reference_lua/checks.rst [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+
+
+.. _checks_string_type_qualifier:
+
+String type qualifier
+---------------------
+
+This section describes how to check a specific argument type using a string type qualifier:
+
+- The :ref:`Supported types <checks_string_supported_types>` section describes all the types supported by the ``checks`` module.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst:91]: This list lacks consistency, which makes it hard to read. Maybe rephrase?

File: doc/reference/reference_lua/checks.rst [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+
+Supported types
+~~~~~~~~~~~~~~~
+
+.. _checks_lua_types:
+
+Lua types
+*********
+
+A string type qualifier can accept any of the `Lua types <https://www.lua.org/pil/2.html>`_, for example, ``string``, ``number``, ``table``, or ``nil``.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst:106]: Why not just "can be any Lua type"? It is easier to read.

File: doc/reference/reference_lua/checks.rst [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+    ..  list-table::
+        :widths: 30 40 30
+        :header-rows: 1
+
+        *   -   Check
+            -   Description
+            -   See also
+
+        *   -   ``checks('datetime')``
+            -   Check whether the specified value is :ref:`datetime_object <datetime_obj>`
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst:141]: I don't see much point in repeating the same sentence in all rows changing only the type (which is obvious from the first column). Can we restructure the table somehow?

File: doc/reference/reference_lua/checks.rst [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+
+
+
+.. _checks_custom_function:
+
+Custom function
+***************
+
+A string type qualifier can accept the name of a custom function that performs arbitrary validations.
+To achieve this, the function should be added to the :ref:`checkers <checks-checkers>` table and should return ``true`` if the value is valid.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: Maybe rephrase to direct instruction: "write a function and add to the table"?

File: doc/reference/reference_lua/checks.rst [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+As for a specific type, you can make a :ref:`union type <checks_union_type>` value optional: ``?number|string``.
+
+
+.. _checks_any_type:
+
+Turning off argument checking
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: Turn off > Skip?

File: doc/reference/reference_lua/checks.rst [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+
+
+
+
+.. _checks_table_type_qualifiers:
+
+Table type qualifier
+--------------------
+
+The type qualifier may be a table. In this case, the following checks are made:
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: This description wasn't enough for me to understand what this is used for. Maybe write that it's for cases when arguments are tables?

File: doc/reference/reference_lua/checks.rst [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+In the next example, the same checks are made for the specified keys.
+
+..  literalinclude:: /code_snippets/test/checks/checks_table_keys_test.lua
+    :language: lua
+    :end-before: -- Tests
+    :dedent:
+
+.. NOTE::
+
+    A table type qualifier may be nested and use tables, too.
```
Comment (p7nov) [inline doc/reference/reference_lua/checks.rst]: "Table qualifiers can be nested"?

**Review by p7nov — APPROVED** [review]
Generally OK, but some places seem hard to read; I suggested simpler/shorter wordings.

File: doc/reference/reference_lua/checks.rst (hunk 7 above)
Comment (DifferentialOrange) [reply doc/reference/reference_lua/checks.rst]: > I actually don't get how our devs decide which modules need require and which don't

Someone just decides that's `rawset(_G, 'checks', checks)` in the module is good idea five years ago and then no one breaks it since they don't have a reason to do so. Actually, in case of repo-installed `checks` you also don't need to require it more than once. Once you have a require in any of your scripts, it is available in any code called later. It's just that built-in modules are required inside the core Tarantool code, so the first require already had happened for any application developer.

Actually, I prefer calling `local checks = require('checks')` in any script, but, well, it is a fact that it is not actually needed. My original comment was related to `checkers` -- it is available in global after first `require('checks')` (which had happen already internally), but you cannot do `require('checkers')` -- it is only in global.

File: doc/reference/reference_lua/checks.rst (hunk 9 above)
Comment (DifferentialOrange) [reply doc/reference/reference_lua/checks.rst]: > Do I get it right that `?` is for checking arbitrary subsets of arguments? For example, in `f(a, b, c, d)`, check only `a` and `d`? If yes, then maybe show an example?

Input count must match with checks expected count (excluding varargs).

```lua
tarantool> g = function(a, b) checks('number', 'number') end
---
...

tarantool> g(1, 2)
---
...

tarantool> g(1)
---
- error: '[string "g = function(a, b) checks(''number'', ''number'')..."]:1: bad argument
    #2 to nil (number expected, got nil)'
... 

tarantool> g = function(a, b) checks('number') end
---
...

tarantool> g(1, 2)
---
- error: '[string "g = function(a, b) checks(''number'') end"]:1: checks: argument
    "b" is not checked'
...

tarantool> g = function(a, b) checks('number', '?') end
---
...

tarantool> g(1)
---
...

```

So, in case you want to check some args and you do not care about other args, you must check them with `'?'` since you cannot simply skip them. It makes even more sence in case you don't want to check some argument in the middle. It is similar for table fields -- if you want to allow optional field X with arbitrary value, you use `'?'`.

File: doc/reference/reference_lua/checks.rst (hunk 10 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst:91]: Consistency is good for lists with more 'homogeneous' items. This list shows different features of string qualifiers and I think it's okay to describe them more like a regular text but using a list for convenience.

File: doc/reference/reference_lua/checks.rst (hunk 11 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst:106]: Strictly speaking, `type qualifier != type`. A type qualifier can accept one type, several types, the `?` value. `Accept` sounds more technically correct here.

File: doc/reference/reference_lua/checks.rst (hunk 12 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst:141]: Auto-generated lists of  API members often look like this. I think, it's not a problem.

File: doc/reference/reference_lua/checks.rst (hunk 14 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst]: Not sure am I right or not but to me `skip` has a slightly different meaning and suits better for cases when we have some kind of a loop. Here, we turn off checking of the specified argument.

File: doc/reference/reference_lua/checks.rst (hunk 15 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst]: Replaced with a sentence from intro.

File: doc/reference/reference_lua/checks.rst (hunk 14 above)
Comment (p7nov) [reply doc/reference/reference_lua/checks.rst]: I see it this way: 
- by default, no args are checked - checking is _turned off_ for all of them.
- _turn off_ checking means return the default (no check) after _turning on_ checking.
- but this text tells about the case when nobody turned on checking for these args :) 

"skip" came to me because I see args as a sequence. So I read a `check` call with a `?` like this: check `arg1`, check `arg2`, skip `arg3`, check `arg4`.

File: doc/reference/reference_lua/checks.rst (hunk 9 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst]: > Do I get it right that ? is for checking arbitrary subsets of arguments? For example, in f(a, b, c, d), check only a and d? If yes, then maybe show an example?

You are right. Will update the example to make it more clear.

File: doc/reference/reference_lua/checks.rst (hunk 14 above)
Comment (andreyaksenov) [reply doc/reference/reference_lua/checks.rst]: Now I got this :) Will update
